Solutions/Corvus.Storage.Azure.Cosmos/Corvus/Storage/Azure/Cosmos/CosmosContainerFactory.cs

```diff
@@ -0,0 +1,138 @@
+// </copyright>
+
+using Azure.Core;
+
+using Corvus.Identity.ClientAuthentication.Azure;
+
+using Microsoft.Azure.Cosmos;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Corvus.Storage.Azure.Cosmos
+{
+    /// <summary>
+    /// A factory for a <see cref="Container"/>.
+    /// </summary>
+    internal class CosmosContainerFactory :
+        TwoLevelCachingStorageContextFactory<CosmosClient, Container, CosmosContainerConfiguration, CosmosClientOptions>,
+        ICosmosContainerSourceByConfiguration
+    {
+        private readonly IAzureTokenCredentialSourceFromDynamicConfiguration azureTokenCredentialSource;
+        private readonly IServiceProvider serviceProvider;
+
+        /// <summary>
+        /// Creates a <see cref="CosmosContainerFactory"/>.
+        /// </summary>
+        /// <param name="azureTokenCredentialSource">
+        /// Provides <see cref="TokenCredential"/>s in exchange for
+        /// <see cref="ClientIdentityConfiguration"/>s.
+        /// </param>
+        /// <param name="serviceProvider">
+        /// Provides access to dependencies that are only needed in certain scenarios, and which
+        /// we don't want to cause a DI initialization failure for if they are absent. (We depend
+        /// on <see cref="IServiceIdentityAzureTokenCredentialSource"/>, but only in certain
+        /// scenarios.)
+        /// </param>
+        public CosmosContainerFactory(
+            IAzureTokenCredentialSourceFromDynamicConfiguration azureTokenCredentialSource,
+            IServiceProvider serviceProvider)
+        {
+            this.azureTokenCredentialSource = azureTokenCredentialSource;
+            this.serviceProvider = serviceProvider;
+        }
+
+        /// <inheritdoc/>
+        protected override ValueTask<Container> CreateContextAsync(
+            CosmosClient parent, CosmosContainerConfiguration configuration, CosmosClientOptions? connectionOptions)
+        {
+            Database db = parent.GetDatabase(configuration.Database);
+            return new ValueTask<Container>(db.GetContainer(configuration.Container));
+        }
+
+        /// <inheritdoc/>
+        protected override async ValueTask<CosmosClient> CreateParentContextAsync(
+            CosmosContainerConfiguration configuration, CosmosClientOptions? clientOptions)
+        {
+            CosmosClient client = await this.CreateCosmosClientAsync(configuration, clientOptions)
+                .ConfigureAwait(false);
+
+            return client;
+        }
+
+        /// <inheritdoc/>
+        protected override string GetCacheKeyForContext(CosmosContainerConfiguration contextConfiguration)
+        {
+            // TODO: there are many options for configuration, and we need to work out a sound way
+            // to reduce that reliably to a cache key.
+            // This is a placeholder that kind of works but is bad for a lot of reasons
+            return System.Text.Json.JsonSerializer.Serialize(contextConfiguration);
```

Comment:
Should we reference the issue that encaspulates "doing something about it" in this comment?

Reply:
Yes, I should have done that. Fixed here and also the the Blob factory which has the same issue.